### PR TITLE
Initial commit of ZenDesk attachment backup script.

### PR DESCRIPTION
**Pull request by ewels**

### CODE REVIEW

File: zendesk_attachment_backup.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                        help="Path to ZenDesk tickets.xml export.")
+   kwargs = vars(parser.parse_args())
+
+   # Get the URLs
+   downloads = collect_urls(**kwargs)
+
+   # Download the URLs
+   if len(downloads) == 0:
+      print("Error - no download URLs found in {}".format(kwargs['xml_path']))
+   elif not kwargs['print_urls']:
```
Comment (senthil10) [inline zendesk_attachment_backup.py]: So it either prints or downloads ?

**Comment by pekrau** [comment]
Line 26: The XML file is read in one go. Not scalable. The file is going to increase in size with time, so why not read it line-by-line and do the matching on lines?

File: zendesk_attachment_backup.py (hunk 1 above)
Comment (ewels) [inline zendesk_attachment_backup.py]: Yup :+1:

**Comment by ewels** [comment]
@pekrau: I was thinking that one global regex search would be better (faster) than many regexes on each line. Also, XML doesn't care about whitespace, so there could be strange linebreaks (or the entire file could be on one line). Granted, it isn't, but I didn't know that when I wrote the code :)

Do you think that the XML will grow to such a size that it won't fit in memory?

**Comment by ewels** [comment]
I should probably mention - the _"proper"_ way to do this of course is to parse the XML. Then we could get extra data about each attachment, such as ticket ID, date and so on. This was a quick and dirty script to just get something in place so I avoided that stuff. Also it could be argued that a simpler approach is more robust to changes in the ZenDesk XML format.

**Comment by ewels** [comment]
I just thought of one update which could be useful - checking that we don't already have a file of that filename before downloading. This would make subsequent backups much faster. However, not sure how they will be run - will it be on a blank location every time?

**Comment by pekrau** [comment]
I don't think XML allows whitespace breaks in string value of type URL? So that is not a concern.

The file being one line: Could happen. But then the Zendesk programmers would be idiots...

XML file size: Well, the size will keep growing indefinitely, but maybe not faster than memory gets cheaper... But I think in general it is good to assume that a file one doesn't know the size of could become huge, and use the appropriate programming pattern. But that's me being old-fashioned and paranoid. I remember the days when 2 MB in RAM was considered utopian...

Proper parsing: I thought about that also, but I don't think it's worth the hassle, since your pattern will handle it.

**Comment by pekrau** [comment]
Checking if file exists: Yes, please do!

Do you mean blank target location (disk on our machine)? Not sure. Maybe we should check with what IT Support considers to be a good policy, but why not just dump it all into one directory, which will grow with new files for each download?

**Comment by ewels** [comment]
@pekrau: Skipping existing files done. Mind if I'm lazy about the other point? I'll volunteer to re-write if it becomes a problem :grinning:

**Comment by pekrau** [comment]
OK, OK. I will hunt you down when the script goes haywire in 20 years time...

**Comment by ewels** [comment]
Deal :) Thanks!

**Pull request merged**